lib/pkce.js

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright (c) 2015-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ *
+ */
+
+ /* eslint-disable complexity, max-statements */
+var http          = require('./http');
+var util          = require('./util');
+var oauthUtil     = require('./oauthUtil');
+var AuthSdkError  = require('./errors/AuthSdkError');
+var token      = require('./token');
+
+// Code verifier: Random URL-safe string with a minimum length of 43 characters.
+// Code challenge: Base64 URL-encoded SHA-256 hash of the code verifier.
+var MIN_VERIFIER_LENGTH = 43;
+function generateVerifier(prefix) {
+  var verifier = prefix || '';
+  if (verifier.length < MIN_VERIFIER_LENGTH) {
+    verifier = verifier + util.genRandomString(MIN_VERIFIER_LENGTH - verifier.length);
+  }
+  return encodeURIComponent(verifier);
+}
+
+/* global Uint8Array, TextEncoder */
+function computeChallenge(str) {  
+  var buffer = new TextEncoder().encode(str);
+  return crypto.subtle.digest('SHA-256', buffer).then(function(arrayBuffer) {
```

Comment:
Since [`crypto.subtle`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle) isn't supported across all of [Okta's supported browsers](https://help.okta.com/en/prod/Content/Topics/Miscellaneous/Platforms_Browser_OS_Support.htm), we're going to want to make sure this get polyfilled or we return a failure state if `crypto.subtle` isn't defined.

A great way to assert this is to run our new e2e test with our existing embedded browser tests. @vijetmahabaleshwar-okta  can you help with this?

Reply:
Would we want to polyfill crypto, or fallback to implicit flow?

Reply:
👍 I will add a method to test for PKCE support. Users could test and set grantType before calling getWithRedirect() or catch the exception and call again with grantType implicit, or just fail. Unless its on an HTTP connection, it should only fail on a few old browsers.